Tensile/Source/client/source/Reference.cpp

```diff
@@ -118,6 +118,42 @@ namespace Tensile
             throw std::runtime_error(msg.c_str());
         }
 
+        /* high precision accumulate: dist type != Accumulator or src type != Accumulator, such as H/H/H + HPA, H/H/S + HPA, B/B/S + HPA, B/S/S + HPA
+         * Ignore types: I8/I32/I32, I8x4/I32/I32
+         */
+        template<typename Inputs, typename Accumulator>
+        constexpr bool is_hpa = !(std::is_same<typename Inputs::AType, Accumulator>() && std::is_same<typename Inputs::BType, Accumulator>() && std::is_same<typename Inputs::DType, Accumulator>())
+                                && !(std::is_same<typename Inputs::AType, Int8>() && std::is_same<typename Inputs::DType, int32_t>())    //case I8/I32/I32
+                                && !(std::is_same<typename Inputs::AType, Int8x4>() && std::is_same<typename Inputs::DType, int32_t>()); //case I8x4/I32/I32
+
+        /* is_hpa and (A!=B or A!=Accumlator or B!=Accumlator) */
+        template<typename A, typename B, typename Inputs, typename Accumulator>
+        constexpr bool need_transform = is_hpa<Inputs, Accumulator> && !(std::is_same<A, Accumulator>() && std::is_same<B, Accumulator>() && std::is_same<A, B>());
+
+        template <typename A, typename B, typename Inputs, typename Accumulator>
+        Accumulator multiply(A a, B b, typename std::enable_if<!need_transform<A, B, Inputs, Accumulator>>::type* = nullptr)
```

Comment:
Replacing `enable_if` with a `bool` reduces code verbosity and remove the need for the template parameter `Input` as your `enable_if` requires it.

Bonus: rearranging `typename Accumulator` to the leftmost side, the rest of them can even be deduced from the function arguments.

```c++
// declaration
template<typename Accumulator, typename A, typename B>
Accumulator multiply(A a, B b, bool need_transform);

// user code
//multiply<typename Inputs::AType, typename Inputs::BType, Inputs, Accumulator>(aVal, bVal);
multiply<Accumulator>(a, b, true);  // need_transform
multiply<Accumulator>(a, b, false); // non need_transform
```

Reply:
@rosenrodt 
Thanks for your advice, however using bool at the function parameter would be another implement way.
The idea here is to avoid the check at runtime, so use enable_if here to use the template mechanism to let the compiler does the check things. Another reason is because of the function template doesn't allow partial specification so, using the enable_if to generate functions is one way to make the idea happened.